x/babylon/types/bsn_contracts.go

```diff
@@ -0,0 +1,44 @@
+package types
+
+import (
+	errorsmod "cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// Equal compares two BSNContracts for equality.
+func (c *BSNContracts) Equal(other *BSNContracts) bool {
+	if c == nil && other == nil {
+		return true
+	}
+	if c == nil || other == nil {
+		return false
+	}
+	return c.BabylonContract == other.BabylonContract &&
+		c.BtcLightClientContract == other.BtcLightClientContract &&
+		c.BtcStakingContract == other.BtcStakingContract &&
+		c.BtcFinalityContract == other.BtcFinalityContract
+}
+
+// ValidateBasic validates the BSNContracts object
+func (c *BSNContracts) ValidateBasic() error {
+	if _, err := sdk.AccAddressFromBech32(c.BabylonContract); err != nil {
+		return errorsmod.Wrap(err, "babylon contract")
+	}
+	if _, err := sdk.AccAddressFromBech32(c.BtcLightClientContract); err != nil {
+		return errorsmod.Wrap(err, "btc light client contract")
+	}
+	if _, err := sdk.AccAddressFromBech32(c.BtcStakingContract); err != nil {
+		return errorsmod.Wrap(err, "btc staking contract")
+	}
+	if _, err := sdk.AccAddressFromBech32(c.BtcFinalityContract); err != nil {
+		return errorsmod.Wrap(err, "btc finality contract")
+	}
+	return nil
+}
+
+func (c *BSNContracts) IsSet() bool {
+	return c.BabylonContract != "" &&
+		c.BtcFinalityContract != "" &&
+		c.BtcLightClientContract != "" &&
+		c.BtcStakingContract != ""
+}
```

Comment:
`ValidateBasic` is superset of this function so is having `ValidateBasic` good enough?

Reply:
That's what I thought initially. But it happens that the object can be set (i.e. not be `nil`) without the contracts inside being set. For example, in block 0 when the contracts are not set, the object is still not `nil`, but the contract strings are empty. This function was added later as an additional condition to the nil check. Not sure if there's a better way to do this.